Currency: Common currency reference conversion

sigma-software-prebid opened the pull request:
### 🔧 Type of changes
- [ ] new bid adapter
- [ ] update bid adapter
- [x] new feature
- [ ] new analytics adapter
- [ ] new module
- [ ] bugfix
- [ ] documentation
- [ ] configuration
- [ ] technical debt (test coverage, refactoring, etc.)

### ✨ What's the context?
https://github.com/prebid/prebid-server/issues/3566

### 🧠 Rationale behind the change
The Prebid.js currency module is able to cross convert currencies using USD (or GBP) as a common currency reference. This feature is not currently implemented in PBS Go, so the following error exists:
`Unable to convert from currency NOK to desired ad server currency SEK`.

In this PR we implement the common currency reference conversion logic (`FindIntermediateConversionRate` method) in the `currency` package.

[Related change in PBS Java](https://github.com/prebid/prebid-server-java/pull/3084)

### 🧪 Test plan
New functionality is covered by `TestGetRate_FindIntermediateConversionRate` unit test in `currency/rates_test.go`.

guscarreon commented on currency/rates.go:
```diff
@@ -46,6 +46,19 @@ func (r *Rates) GetRate(from, to string) (float64, error) {
+			if fromPresent {
+				return toConversion / fromConversion, nil
+			}
+		}
```
Can we move lines 51 to 60 to their own function and code a table-driven test in `currency/rates_test.go`?

sigma-software-prebid replied:
@guscarreon our intention was to fulfil the feature [request](https://github.com/prebid/prebid-server/issues/3566) where it has been mentioned as an `intermediate currency`.

Moreover, we have investigated the implementation in [PBS Java](https://github.com/prebid/prebid-server-java/pull/3084/files) and [Prebid.js](https://github.com/prebid/Prebid.js/blob/882a9920700d5435a28cfe6e46be98c6d3c8a0af/modules/currency.js#L276) and based on the findings we wanted to be as close to "spec" as possible in order not to run into inconsistency between implementations - Go and Java versions being nearly identical and Prebid.js with minimal differences. However, I agree that `common currency reference conversion` name would fit better.

Also, I guess if we want to make more functional implementation it would be much better to extend conversion tables either to `n*n` currency mapping to minimize runtime efforts (plain lookup instead of looping) or make all the logic via one explicit common currency, but then this issue would have a bit wider scope and all 3 implementation between different Prebid versions would have different logic implemented.

What do you think?

guscarreon reviewed (commented):
Take the conversion rates shown below:
```
	rates := NewRates(map[string]map[string]float64{
		"USD": {
		    "CAD": 0.90,
		    "MXN": 20.00,
		    "JPY": 50.00,
		},
		"MXN": {
		    "EUR": 0.05,
		},
	})
```
MAybe it's just me but, from "intermediate conversion" I would think that I'd be able to convert `USD` to `EUR` via `MXN`. In other words, convert from `USD` to `MXN` first, followed by a second conversion from `MXN` to `EUR` (`USD` -> `MXN` -> `EUR`) but, with the logic implemented in this pull request, that's not the case, right? The implemented logic would help me say, convert from Canadian Dollars to Japanese Yens (`CAD` -> `JPY`) that are both found under `USD`.
1. Do you intend to implement `USD` -> `MXN` -> `EUR` conversions too?
1. If not, should we rename this type of conversion to "common currency reference conversion" or something of sorts? Or do you agree with the naming?

sigma-software-prebid commented:
> Take the conversion rates shown below:
> 
> ```
> 	rates := NewRates(map[string]map[string]float64{
> 		"USD": {
> 		    "CAD": 0.90,
> 		    "MXN": 20.00,
> 		    "JPY": 50.00,
> 		},
> 		"MXN": {
> 		    "EUR": 0.05,
> 		},
> 	})
> ```
> 
> MAybe it's just me but, from "intermediate conversion" I would think that I'd be able to convert `USD` to `EUR` via `MXN`. In other words, convert from `USD` to `MXN` first, followed by a second conversion from `MXN` to `EUR` (`USD` -> `MXN` -> `EUR`) but, with the logic implemented in this pull request, that's not the case, right? The implemented logic would help me say, convert from Canadian Dollars to Japanese Yens (`CAD` -> `JPY`) that are both found under `USD`.
> 
> 1. Do you intend to implement `USD` -> `MXN` -> `EUR` conversions too?
> 2. If not, should we rename this type of conversion to "common currency reference conversion" or something of sorts? Or do you agree with the naming?

@guscarreon our intention was to fulfil the feature https://github.com/prebid/prebid-server/issues/3566 where it has been mentioned as an intermediate currency.

Moreover, we have investigated the implementation in [PBS Java](https://github.com/prebid/prebid-server-java/pull/3084/files) and [Prebid.js](https://github.com/prebid/Prebid.js/blob/882a9920700d5435a28cfe6e46be98c6d3c8a0af/modules/currency.js#L276) and based on the findings we wanted to be as close to "spec" as possible in order not to run into inconsistency between implementations - Go and Java versions being nearly identical and Prebid.js with minimal differences. However, I agree that common currency reference conversion name would fit better.

Also, I guess if we want to make more functional implementation it would be much better to extend conversion tables either to n*n currency mapping to minimize runtime efforts (plain lookup instead of looping) or make all the logic via one explicit common currency, but then this issue would have a bit wider scope and all 3 implementation between different Prebid versions would have different logic implemented.

What do you think?

guscarreon commented:
> > Take the conversion rates shown below:
> > ```
> > 	rates := NewRates(map[string]map[string]float64{
> > 		"USD": {
> > 		    "CAD": 0.90,
> > 		    "MXN": 20.00,
> > 		    "JPY": 50.00,
> > 		},
> > 		"MXN": {
> > 		    "EUR": 0.05,
> > 		},
> > 	})
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > MAybe it's just me but, from "intermediate conversion" I would think that I'd be able to convert `USD` to `EUR` via `MXN`. In other words, convert from `USD` to `MXN` first, followed by a second conversion from `MXN` to `EUR` (`USD` -> `MXN` -> `EUR`) but, with the logic implemented in this pull request, that's not the case, right? The implemented logic would help me say, convert from Canadian Dollars to Japanese Yens (`CAD` -> `JPY`) that are both found under `USD`.
> > 
> > 1. Do you intend to implement `USD` -> `MXN` -> `EUR` conversions too?
> > 2. If not, should we rename this type of conversion to "common currency reference conversion" or something of sorts? Or do you agree with the naming?
> 
> @guscarreon our intention was to fulfil the feature #3566 where it has been mentioned as an intermediate currency.
> 
> Moreover, we have investigated the implementation in [PBS Java](https://github.com/prebid/prebid-server-java/pull/3084/files) and [Prebid.js](https://github.com/prebid/Prebid.js/blob/882a9920700d5435a28cfe6e46be98c6d3c8a0af/modules/currency.js#L276) and based on the findings we wanted to be as close to "spec" as possible in order not to run into inconsistency between implementations - Go and Java versions being nearly identical and Prebid.js with minimal differences. However, I agree that common currency reference conversion name would fit better.
> 
> Also, I guess if we want to make more functional implementation it would be much better to extend conversion tables either to n*n currency mapping to minimize runtime efforts (plain lookup instead of looping) or make all the logic via one explicit common currency, but then this issue would have a bit wider scope and all 3 implementation between different Prebid versions would have different logic implemented.
> 
> What do you think?
 
@sigma-software-prebid thank you for the clarification. I agree with catching up to PBS-Java and Prebid.js. One final question: can we, as stated in a comment on my previous review above, move lines 51 to 60 in `currency/rates.go` to their own function and code a table-driven test in `currency/rates_test.go`? Let me know if you agree with this approach

sigma-software-prebid commented:
> > > Take the conversion rates shown below:
> > > ```
> > > 	rates := NewRates(map[string]map[string]float64{
> > > 		"USD": {
> > > 		    "CAD": 0.90,
> > > 		    "MXN": 20.00,
> > > 		    "JPY": 50.00,
> > > 		},
> > > 		"MXN": {
> > > 		    "EUR": 0.05,
> > > 		},
> > > 	})
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > MAybe it's just me but, from "intermediate conversion" I would think that I'd be able to convert `USD` to `EUR` via `MXN`. In other words, convert from `USD` to `MXN` first, followed by a second conversion from `MXN` to `EUR` (`USD` -> `MXN` -> `EUR`) but, with the logic implemented in this pull request, that's not the case, right? The implemented logic would help me say, convert from Canadian Dollars to Japanese Yens (`CAD` -> `JPY`) that are both found under `USD`.
> > > 
> > > 1. Do you intend to implement `USD` -> `MXN` -> `EUR` conversions too?
> > > 2. If not, should we rename this type of conversion to "common currency reference conversion" or something of sorts? Or do you agree with the naming?
> > 
> > 
> > @guscarreon our intention was to fulfil the feature #3566 where it has been mentioned as an intermediate currency.
> > Moreover, we have investigated the implementation in [PBS Java](https://github.com/prebid/prebid-server-java/pull/3084/files) and [Prebid.js](https://github.com/prebid/Prebid.js/blob/882a9920700d5435a28cfe6e46be98c6d3c8a0af/modules/currency.js#L276) and based on the findings we wanted to be as close to "spec" as possible in order not to run into inconsistency between implementations - Go and Java versions being nearly identical and Prebid.js with minimal differences. However, I agree that common currency reference conversion name would fit better.
> > Also, I guess if we want to make more functional implementation it would be much better to extend conversion tables either to n*n currency mapping to minimize runtime efforts (plain lookup instead of looping) or make all the logic via one explicit common currency, but then this issue would have a bit wider scope and all 3 implementation between different Prebid versions would have different logic implemented.
> > What do you think?
> 
> @sigma-software-prebid thank you for the clarification. I agree with catching up to PBS-Java and Prebid.js. One final question: can we, as stated in a comment on my previous review above, move lines 51 to 60 in `currency/rates.go` to their own function and code a table-driven test in `currency/rates_test.go`? Let me know if you agree with this approach

@guscarreon thank you for the feedback. We have extracted the above-mentioned code into the separate `FindConversionRate` method and adjusted the PR title accordingly.

guscarreon reviewed (commented):
@sigma-software-prebid thank you for addressing my feedback. LGTM

sigma-software-prebid commented on currency/rates_test.go:
```diff
@@ -186,6 +187,87 @@ func TestGetRate_ReverseConversion(t *testing.T) {
 	}
 }
 
+func TestGetRate_FindConversionRate(t *testing.T) {
```
@bsardo thank you for the review. We have addressed all your feedback. Please, let us know what do you think.
